src/main/java/net/neoforged/neoform/runtime/downloads/ParallelDownloader.java

```diff
@@ -0,0 +1,149 @@
+package net.neoforged.neoform.runtime.downloads;
+
+import net.neoforged.neoform.runtime.utils.FileUtil;
+import net.neoforged.neoform.runtime.utils.Logger;
+import net.neoforged.neoform.runtime.utils.StringUtil;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class is capable of download a large number of files concurrently, while observing a maximum
+ * concurrent download limit.
+ */
+public class ParallelDownloader implements AutoCloseable {
+    private static final Logger LOG = Logger.create();
+    private static final ThreadFactory DOWNLOAD_THREAD_FACTORY = r -> Thread.ofVirtual().name("parallel-download", 1).unstarted(r);
```

Comment:
You can use `factory` on the thread builder. `unstarted` is pointless since you're wrapping it in an executor.